### ISSUE TITLE
Retrofit 2.0.0-beta2 android.os.NetworkOnMainThreadException

### ISSUE THREAD

**Issue by seiton-it** [JakeWharton/u2020#215]
Hi there is some serious bug in that retrofit version. 
i get exception 
```
Failed to get trending repositories
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView: android.os.NetworkOnMainThreadException
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1147)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at java.net.InetAddress.lookupHostByName(InetAddress.java:418)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at java.net.InetAddress.getAllByNameImpl(InetAddress.java:252)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at java.net.InetAddress.getAllByName(InetAddress.java:215)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.Network$1.resolveInetAddresses(Network.java:29)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:187)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:156)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.RouteSelector.next(RouteSelector.java:98)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.HttpEngine.createNextConnection(HttpEngine.java:344)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:327)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:245)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call.getResponse(Call.java:267)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:224)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.jakewharton.u2020.data.api.LoggingInterceptor.intercept(LoggingInterceptor.java:28)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:221)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.jakewharton.u2020.data.api.oauth.OauthInterceptor.intercept(OauthInterceptor.java:25)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:221)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:195)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.squareup.okhttp.Call.execute(Call.java:79)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at retrofit.OkHttpCall.execute(OkHttpCall.java:116)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at retrofit.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:111)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at retrofit.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:88)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.Observable$2.call(Observable.java:162)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.Observable$2.call(Observable.java:154)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.Observable$2.call(Observable.java:162)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.Observable$2.call(Observable.java:154)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.Observable.unsafeSubscribe(Observable.java:7710)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:231)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:140)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.internal.operators.OperatorMap$1.onNext(OperatorMap.java:55)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.subjects.SubjectSubscriptionManager$SubjectObserver.onNext(SubjectSubscriptionManager.java:224)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at rx.subjects.PublishSubject.onNext(PublishSubject.java:114)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.jakewharton.u2020.ui.trending.TrendingView$5.run(TrendingView.java:182)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at android.os.Handler.handleCallback(Handler.java:739)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at android.os.Handler.dispatchMessage(Handler.java:95)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at android.os.Looper.loop(Looper.java:135)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at android.app.ActivityThread.main(ActivityThread.java:5312)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at java.lang.reflect.Method.invoke(Native Method)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at java.lang.reflect.Method.invoke(Method.java:372)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:901)
10-12 21:29:04.583 28473-28473/com.jakewharton.u2020.internal.debug E/TrendingView:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:696)
```
If i switch to retrofit 2.0.0-beta1 everything works like charm

**Comment by mdrabic** [comment]
@seiton-it Can you provide a snippet of your observable sequence? 

If your chain of observables looks like something below, make sure you apply the `subscribeOn()` operator to any additional observables created in the middle of the sequence. ie, the line `details.add(myApi.getDetails(h.getId()).subscribeOn(Schedulers.io()));`.  Not applying the second `subscribeOn()` operator resulted in a `NetworkOnMainThreadException`. 

```java
myApi.getHeadlines()
    .subscribeOn(Schedulers.io())
    .observeOn(AndroidSchedulers.mainThread())
    .flatMap(new Func1<Headlines, Observable<HeadlineDetail> {
        @Override
        public Observable<HeadlineDetail> call(Headlines headlines) {
            List<Observable<HeadlineDetail>> details = new ArrayList<>();
            for (Headline h : headlines) {
                details.add(myApi.getDetails(h.getId()).subscribeOn(Schedulers.io()));
            }
            return Observable.merge(details);
        }
    })
    .subscribe(...);
```

I would like to know if it is expected that the `onSubscribe()` operator would need to be applied in this situation. Can anyone comment on that? My initial thought is since the `onSubscribe()` operator was already applied in the sequence, any additional observables in the sequence would have that applied to it as well.

**Comment by JakeWharton** [comment]
Retrofit 2's observables execute synchronously. Use `subscribeOn` to put its work on a background thread.

**Comment by p-fischer** [comment]
I observed similar behavior. I'm performing a call from main thread using the *asynchronous* option with *enqueue()*.
Additionally I added this interceptor

```
public class MyInterceptor implements com.squareup.okhttp.Interceptor
{
	@Override
	public Response intercept( Chain chain ) throws IOException
	{
		Request request = chain.request();

		Log.i( "Interceptor", "before request" );

		Response response = chain.proceed( request );

		Log.i( "Interceptor", "after request" );

		return response;
	}
}
```
... with this line: `retrofit.client().interceptors().add( new MyInterceptor() );`

I would expect the interceptor to run in an asynchronous thread, e.g. the one which is created for my asynchronous call. 
With *retrofit 2.0.0-beta1* my code works. However with *retrofit 2.0.0-beta2* I get the *NetworkOnMainThreadException* exception.

Is this a bug or an intential change? How do I fix my interceptor?
